File: desktop/src/utils/harness/schedule.ts
/**
 * Schedule helpers shared by the Harness Scheduled view and the cron proxy.
 * Cron expressions match the Hermes `cronjob` runtime on the VPS.
 */

import type {
  HarnessSchedule,
  HarnessScheduleTemplate,
  HarnessWeekday,
} from '@/types/harness'

/** Weekday keys in display order (Monday first). */
export const HARNESS_WEEKDAYS: readonly HarnessWeekday[] = [
  'mon',
  'tue',
  'wed',
  'thu',
  'fri',
  'sat',
  'sun',
]

/** Weekday key to cron day-of-week number (Sunday is 0). */
const CRON_DAY: Record<HarnessWeekday, number> = {
  sun: 0,
  mon: 1,
  tue: 2,
  wed: 3,
  thu: 4,
  fri: 5,
  sat: 6,
}

/** Built-in office templates offered in the create form. */
export const HARNESS_SCHEDULE_TEMPLATES: readonly HarnessScheduleTemplate[] = [
  { id: 'dailyBrief', schedule: { kind: 'weekdays', time: '08:00', days: [] }, target: 'vps' },
  { id: 'weeklyReview', schedule: { kind: 'weekly', time: '16:00', days: ['fri'] }, target: 'vps' },
  {
    id: 'followUpMonitor',
    schedule: { kind: 'weekdays', time: '09:00', days: [] },
    target: 'vps',
  },
]

/**
 * Splits `HH:MM` into numeric parts, falling back to 09:00.
 * @param time - 24h time string.
 * @returns Hour and minute.
 */
function parseTime(time: string): { hour: number; minute: number } {
  const match = /^(\d{1,2}):(\d{2})$/.exec(time.trim())
  const hour = Number(match?.[1] ?? '9')
  const minute = Number(match?.[2] ?? '0')
  if (!Number.isFinite(hour) || !Number.isFinite(minute) || hour > 23 || minute > 59) {
    return { hour: 9, minute: 0 }
  }
  return { hour, minute }
}

/**
 * Returns whether a value is a valid 24h `HH:MM` string.
 * @param time - Candidate value.
 * @returns True when the string is a usable time.
 */
export function isValidScheduleTime(time: string): boolean {
  const match = /^(\d{1,2}):(\d{2})$/.exec(time.trim())
  if (!match) {
    return false
  }
  return Number(match[1]) <= 23 && Number(match[2]) <= 59
}

/**
 * Converts a schedule to a Hermes cron expression.
 * @param schedule - Recurrence chosen in the UI.
 * @returns Five-field cron string.
 */
export function scheduleToCron(schedule: HarnessSchedule): string {
  const { hour, minute } = parseTime(schedule.time)
  if (schedule.kind === 'daily') {
    return `${minute} ${hour} * * *`
  }
  if (schedule.kind === 'weekdays') {
    return `${minute} ${hour} * * 1-5`
  }
  const days = schedule.days.length > 0 ? schedule.days : (['mon'] as HarnessWeekday[])
  const sorted = [...new Set(days)].map((day) => CRON_DAY[day]).sort((a, b) => a - b)
  return `${minute} ${hour} * * ${sorted.join(',')}`
}

/**
 * Whether a schedule fires on the given weekday.
 * @param schedule - Recurrence to test.
 * @param jsDay - `Date.getDay()` value (Sunday is 0).
 * @returns True when the day is included.
 */
function firesOnDay(schedule: HarnessSchedule, jsDay: number): boolean {
  if (schedule.kind === 'daily') {
    return true
  }
  if (schedule.kind === 'weekdays') {
    return jsDay >= 1 && jsDay <= 5
  }
  const days = schedule.days.length > 0 ? schedule.days : (['mon'] as HarnessWeekday[])
  return days.some((day) => CRON_DAY[day] === jsDay)
}

/**
 * Computes the next local fire time for a schedule.
 * @param schedule - Recurrence to evaluate.
 * @param fromMs - Reference timestamp; defaults to now.
 * @returns Epoch milliseconds of the next run, or null when unreachable.
 */
export function nextRunAtMs(schedule: HarnessSchedule, fromMs: number = Date.now()): number | null {
  const { hour, minute } = parseTime(schedule.time)
  const cursor = new Date(fromMs)
  cursor.setSeconds(0, 0)

  for (let offset = 0; offset < 8; offset += 1) {
    const candidate = new Date(cursor)
    candidate.setDate(cursor.getDate() + offset)
    candidate.setHours(hour, minute, 0, 0)
    if (candidate.getTime() > fromMs && firesOnDay(schedule, candidate.getDay())) {
      return candidate.getTime()
    }
  }
  return null
}

File: backend/internal/harness/schedule.go
package harness

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Schedule is the recurrence the desktop Scheduled view offers. It is kept
// structured so the next-run time (and the five-field expression used in
// tests) are derived from one source.
type Schedule struct {
	// Kind is "daily", "weekdays", or "weekly".
	Kind string `json:"kind"`
	// Time is 24h "HH:MM" in the server's local zone.
	Time string `json:"time"`
	// Days applies to "weekly" only: mon..sun.
	Days []string `json:"days"`
	// TimeZone is an IANA zone supplied by the desktop, for example Asia/Shanghai.
	TimeZone string `json:"timeZone,omitempty"`
}

// cronDay maps a weekday key to its cron day-of-week number (Sunday is 0).
var cronDay = map[string]int{
	"sun": 0,
	"mon": 1,
	"tue": 2,
	"wed": 3,
	"thu": 4,
	"fri": 5,
	"sat": 6,
}

// parseTime splits "HH:MM", falling back to 09:00 for malformed input.
func (s Schedule) parseTime() (int, int) {
	hour, minute, ok := strings.Cut(strings.TrimSpace(s.Time), ":")
	if !ok {
		return 9, 0
	}
	h, errH := strconv.Atoi(hour)
	m, errM := strconv.Atoi(minute)
	if errH != nil || errM != nil || h < 0 || h > 23 || m < 0 || m > 59 {
		return 9, 0
	}
	return h, m
}

// normalizedDays returns the weekly day list, defaulting to Monday.
func (s Schedule) normalizedDays() []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(s.Days))
	for _, day := range s.Days {
		key := strings.ToLower(strings.TrimSpace(day))
		if _, ok := cronDay[key]; ok && !seen[key] {
			seen[key] = true
			out = append(out, key)
		}
	}
	if len(out) == 0 {
		return []string{"mon"}
	}
	return out
}

// Valid reports whether the schedule can be scheduled.
func (s Schedule) Valid() bool {
	switch s.Kind {
	case "daily", "weekdays", "weekly":
	default:
		return false
	}
	hour, minute, ok := strings.Cut(strings.TrimSpace(s.Time), ":")
	if !ok {
		return false
	}
	h, errH := strconv.Atoi(hour)
	m, errM := strconv.Atoi(minute)
	if errH != nil || errM != nil || h < 0 || h > 23 || m < 0 || m > 59 {
		return false
	}
	if strings.TrimSpace(s.TimeZone) != "" {
		_, err := time.LoadLocation(strings.TrimSpace(s.TimeZone))
		return err == nil
	}
	return true
}

// CronExpression renders the five-field expression for the structured schedule.
func (s Schedule) CronExpression() string {
	h, m := s.parseTime()
	switch s.Kind {
	case "daily":
		return fmt.Sprintf("%d %d * * *", m, h)
	case "weekdays":
		return fmt.Sprintf("%d %d * * 1-5", m, h)
	default:
		days := s.normalizedDays()
		numbers := make([]int, 0, len(days))
		for _, day := range days {
			numbers = append(numbers, cronDay[day])
		}
		sort.Ints(numbers)
		parts := make([]string, 0, len(numbers))
		for _, n := range numbers {
			parts = append(parts, strconv.Itoa(n))
		}
		return fmt.Sprintf("%d %d * * %s", m, h, strings.Join(parts, ","))
	}
}

// firesOn reports whether the schedule runs on the given weekday.
func (s Schedule) firesOn(weekday time.Weekday) bool {
	switch s.Kind {
	case "daily":
		return true
	case "weekdays":
		return weekday >= time.Monday && weekday <= time.Friday
	default:
		for _, day := range s.normalizedDays() {
			if cronDay[day] == int(weekday) {
				return true
			}
		}
		return false
	}
}

// NextRunAfter returns the next local fire time strictly after `from`.
func (s Schedule) NextRunAfter(from time.Time) (time.Time, bool) {
	h, m := s.parseTime()
	location := from.Location()
	if zone := strings.TrimSpace(s.TimeZone); zone != "" {
		loaded, err := time.LoadLocation(zone)
		if err != nil {
			return time.Time{}, false
		}
		location = loaded
	}
	localFrom := from.In(location)
	for offset := range 8 {
		day := localFrom.AddDate(0, 0, offset)
		candidate := time.Date(day.Year(), day.Month(), day.Day(), h, m, 0, 0, location)
		if candidate.After(from) && s.firesOn(candidate.Weekday()) {
			return candidate, true
		}
	}
	return time.Time{}, false
}
